File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/perpetual/HyperliquidSubscriptionService.kt
package com.gemwallet.android.data.repositories.perpetual

import android.util.Log
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.channels.ReceiveChannel
import uniffi.gemstone.PerpetualAccountMode as GemPerpetualAccountMode
import uniffi.gemstone.GemPerpetualSubscription
import uniffi.gemstone.HyperliquidSubscriptions

class HyperliquidSubscriptionService(
    private val subscriptions: HyperliquidSubscriptions,
) {
    private val outgoing = Channel<String>(Channel.UNLIMITED)
    val messages: ReceiveChannel<String> = outgoing

    fun subscribe(subscription: GemPerpetualSubscription) {
        send { subscriptions.subscribe(subscription) }
    }

    fun unsubscribe(subscription: GemPerpetualSubscription) {
        send { subscriptions.unsubscribe(subscription) }
    }

    fun connected(address: String, mode: GemPerpetualAccountMode) {
        send { subscriptions.connected(address, mode) }
    }

    fun disconnected() {
        subscriptions.disconnected()
    }

    private fun send(requests: () -> List<String>) {
        runCatching(requests)
            .onSuccess { it.forEach(outgoing::trySend) }
            .onFailure { Log.e(TAG, "Subscription request error", it) }
    }

    companion object {
        private const val TAG = "HyperliquidSubscriptions"
    }
}

File: ios/Packages/FeatureServices/PerpetualService/HyperliquidSubscriptionService.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import enum Gemstone.GemPerpetualSubscription
import class Gemstone.HyperliquidSubscriptions
import Primitives
import WebSocketClient

public actor HyperliquidSubscriptionService {
    private let webSocket: any WebSocketConnectable
    private let subscriptions = HyperliquidSubscriptions()

    public init(webSocket: any WebSocketConnectable) {
        self.webSocket = webSocket
    }

    public func subscribe(_ subscription: GemPerpetualSubscription) async throws {
        try await send(subscriptions.subscribe(subscription: subscription))
    }

    public func unsubscribe(_ subscription: GemPerpetualSubscription) async throws {
        try await send(subscriptions.unsubscribe(subscription: subscription))
    }

    public func connected(address: String, mode: PerpetualAccountMode) async throws {
        try await send(subscriptions.connected(address: address, mode: mode.map()))
    }

    public func disconnected() {
        subscriptions.disconnected()
    }

    // MARK: - Private

    private func send(_ requests: [String]) async throws {
        for request in requests {
            try await webSocket.send(request)
        }
    }
}
